### PR TITLE
task/WG-83: Handle Overlapping Point Clouds

### PR DESCRIPTION
**Pull request by tjgrafft**
## Overview: ##
Putting a point marker with the polygon/point cloud. This would make it be able to be seen if zoomed out and would allow users to select different overlapping point clouds. Adding those point markers to the marker cluster, as well.
## PR Status: ##

* [X] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [WG-83](https://tacc-main.atlassian.net/browse/WG-83)

## Summary of Changes: ##

- Added a custom marker to the top-left corner of point clouds, so that users can click the icon and get the correct modal to pop up in cases where there were overlapping point clouds in the same location.
- Added the ability to attach feature data to the marker options prop (so that the modals open when the markers are clicked, which is consistent/same behavior as clicking inside the polygon area)
- Added function to calculate the top-left corner of a layer/feature (this could be refactored or added onto later for other corners)
- Added point cloud markers to the marker cluster group, so that all assets would be included in clusters.
- Added conditionals to the featureClickHandler function to check whether clicked object was a marker or layer.

## Testing Steps: ##
1. Go to localhost:4200 and find a project where there are overlapping point clouds present. (If you need a project, Nathan or I can share one with you)
2. Go to the overlapping point clouds on the map. Check that the new point-cloud icon is present in the top-left corner of the polygons.
3. Click the point-cloud icons and make sure the correct modals open for that point-cloud asset.
4. Zoom out and make sure you see cluster markers and corresponding numbers related to how many assets are in that area--including the overlapping point clouds.
5. Compare with images below

## UI Photos:
**New Point Cloud Markers**
Point_cloud_icons
<br>
**Cluster markers with point-cloud assets included**
cluster_marker2
cluster_marker1

### CODE REVIEW

**Review by taoteg — APPROVED** [review]
LGTM! Ran it locally and worked as advertised.

Note: to test against the Staging env, you will need to disable lines 41-66 to match this commit in `feature/questionnaire` (until that branch gets merged):

https://github.com/TACC-Cloud/hazmapper/commit/da0f40b1a55fcd16053a567a0dbb812e0a98c1a7#diff-7bedcb7bd6c40fef89e5075610ff1cc53e7d0f64a3fb6f06632d56caa966f499

File: angular/src/app/components/map/map.component.ts [hunk 1]
```diff
@@ -350,6 +350,15 @@ export class MapComponent implements OnInit, OnDestroy {
         let feat: LayerGroup;
         if (d.geometry.type === 'Polygon' && d.properties.style) {
           feat = L.geoJSON(d, { style: d.properties.style });
+        } else if (d.featureType() === 'point_cloud') {
+          feat = L.geoJSON(d, { style: d.properties.style });
+          this.features.addLayer(feat);
```
Comment (nathanfranklin) [inline angular/src/app/components/map/map.component.ts:355]:
```suggestion
```

This should be deleted. 


as later on in the code we add it here
```
        if (d.geometry.type === 'Point') {
          markers.addLayer(feat);
        } else {
          this.features.addLayer(feat);
        }
```

File: angular/src/app/components/map/map.component.ts [hunk 2]
```diff
@@ -350,6 +350,15 @@ export class MapComponent implements OnInit, OnDestroy {
...
+          const markerPosition = calculateMarkerPosition(d.geometry);
+          const pointCloudMarker = createMarker(d, markerPosition);
+          pointCloudMarker.on('click', (ev) => {
+            this.featureClickHandler(ev, 'click');
+          });
+          markers.addLayer(pointCloudMarker);
```
Comment (nathanfranklin) [inline angular/src/app/components/map/map.component.ts:361]: for completeness we should probably add the following:

feat.setZIndex(1);

Comment (nathanfranklin) [reply angular/src/app/components/map/map.component.ts:361]: hmm, not so certain about this so just ignore.

**Review by nathanfranklin — APPROVED** [review]
LGTM 👍

**Pull request merged**